From 36ad061171260437bf9f6e3f0cc743f25d2744c0 Mon Sep 17 00:00:00 2001
From: Ma Wupeng <mawupeng1@huawei.com>
Date: Fri, 11 Nov 2022 11:33:01 +0800
Subject: [PATCH 0015/1398] [Huawei] mm: mem_reliable: Bring total code
 isolation for memory reliable

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187270

--------------------------------

Bring total code isolation for memory reliable.

Signed-off-by: Ma Wupeng <mawupeng1@huawei.com>
---
 include/linux/mem_reliable.h | 17 +++++++++++++++++
 include/linux/memblock.h     |  4 ++++
 mm/hugetlb.c                 | 11 +++++++++++
 mm/khugepaged.c              |  4 ++++
 mm/memblock.c                |  6 ++++++
 mm/page_alloc.c              | 13 +++++++++++++
 mm/shmem.c                   | 16 ----------------
 7 files changed, 55 insertions(+), 16 deletions(-)

diff --git a/include/linux/mem_reliable.h b/include/linux/mem_reliable.h
index ddadf2803742..003be6463237 100644
--- a/include/linux/mem_reliable.h
+++ b/include/linux/mem_reliable.h
@@ -135,6 +135,22 @@ static inline void reliable_page_counter(struct page *page,
 	if (page_reliable(page))
 		atomic_long_add(val, &mm->reliable_nr_page);
 }
+
+static inline bool shmem_prepare_alloc(gfp_t *gfp_mask)
+{
+	if (!shmem_reliable_is_enabled())
+		return true;
+
+	if (mem_reliable_shmem_limit_check()) {
+		*gfp_mask |= GFP_RELIABLE;
+		return true;
+	}
+
+	if (reliable_allow_fb_enabled())
+		return true;
+
+	return false;
+}
 #else
 #define reliable_enabled 0
 #define pagecache_use_reliable_mem 0
@@ -178,6 +194,7 @@ static inline void reliable_page_counter(struct page *page,
 		struct mm_struct *mm, int val) {}
 static inline void reliable_report_usage(struct seq_file *m,
 		struct mm_struct *mm) {}
+static inline bool shmem_prepare_alloc(gfp_t *gfp_mask) { return true; }
 #endif
 
 #endif
diff --git a/include/linux/memblock.h b/include/linux/memblock.h
index d6b6a93aa73e..473e84339704 100644
--- a/include/linux/memblock.h
+++ b/include/linux/memblock.h
@@ -38,7 +38,9 @@ enum memblock_flags {
 	MEMBLOCK_MIRROR		= 0x2,	/* mirrored region */
 	MEMBLOCK_NOMAP		= 0x4,	/* don't add to kernel direct mapping */
 	MEMBLOCK_MEMMAP		= 0x8,	/* memmap reserved region */
+#ifdef CONFIG_MEMORY_RELIABLE
 	MEMBLOCK_NOMIRROR	= 0x10,	/* alloc from non-mirrored region */
+#endif
 };
 
 /**
@@ -411,10 +413,12 @@ void *memblock_alloc_try_nid_raw(phys_addr_t size, phys_addr_t align,
 void *memblock_alloc_try_nid(phys_addr_t size, phys_addr_t align,
 			     phys_addr_t min_addr, phys_addr_t max_addr,
 			     int nid);
+#ifdef CONFIG_MEMORY_RELIABLE
 void *memblock_alloc_try_nid_raw_flags(phys_addr_t size, phys_addr_t align,
 				       phys_addr_t min_addr,
 				       phys_addr_t max_addr, int nid,
 				       enum memblock_flags flags);
+#endif
 
 static inline void * __init memblock_alloc(phys_addr_t size,  phys_addr_t align)
 {
diff --git a/mm/hugetlb.c b/mm/hugetlb.c
index d687f271901e..c5a4792ece6b 100644
--- a/mm/hugetlb.c
+++ b/mm/hugetlb.c
@@ -2697,6 +2697,7 @@ struct page *alloc_huge_page(struct vm_area_struct *vma,
 	return ERR_PTR(-ENOSPC);
 }
 
+#ifdef CONFIG_MEMORY_RELIABLE
 static void *__init __alloc_bootmem_huge_page_inner(phys_addr_t size,
 						    phys_addr_t align,
 						    phys_addr_t min_addr,
@@ -2710,6 +2711,16 @@ static void *__init __alloc_bootmem_huge_page_inner(phys_addr_t size,
 	return memblock_alloc_try_nid_raw_flags(size, align, max_addr, max_addr,
 						nid, MEMBLOCK_NOMIRROR);
 }
+#else
+static inline void *__init __alloc_bootmem_huge_page_inner(phys_addr_t size,
+							   phys_addr_t align,
+							   phys_addr_t min_addr,
+							   phys_addr_t max_addr,
+							   int nid)
+{
+	return memblock_alloc_try_nid_raw(size, align, max_addr, max_addr, nid);
+}
+#endif
 
 int alloc_bootmem_huge_page(struct hstate *h, int nid)
 	__attribute__ ((weak, alias("__alloc_bootmem_huge_page")));
diff --git a/mm/khugepaged.c b/mm/khugepaged.c
index dde44ff560e4..331c31cd36d2 100644
--- a/mm/khugepaged.c
+++ b/mm/khugepaged.c
@@ -1078,8 +1078,10 @@ static void collapse_huge_page(struct mm_struct *mm,
 	/* Only allocate from the target node */
 	gfp = alloc_hugepage_khugepaged_gfpmask() | __GFP_THISNODE;
 
+#ifdef CONFIG_MEMORY_RELIABLE
 	if (reliable)
 		gfp |= GFP_RELIABLE;
+#endif
 
 	/*
 	 * Before allocating the hugepage, release the mmap_lock read lock.
@@ -1665,8 +1667,10 @@ static void collapse_file(struct mm_struct *mm,
 	/* Only allocate from the target node */
 	gfp = alloc_hugepage_khugepaged_gfpmask() | __GFP_THISNODE;
 
+#ifdef CONFIG_MEMORY_RELIABLE
 	if (reliable)
 		gfp |= GFP_RELIABLE;
+#endif
 
 	new_page = khugepaged_alloc_page(hpage, gfp, node);
 	if (!new_page) {
diff --git a/mm/memblock.c b/mm/memblock.c
index e1fd07166a35..3fb02af2c97c 100644
--- a/mm/memblock.c
+++ b/mm/memblock.c
@@ -968,9 +968,11 @@ static bool should_skip_region(struct memblock_type *type,
 	if ((flags & MEMBLOCK_MIRROR) && !memblock_is_mirror(m))
 		return true;
 
+#ifdef CONFIG_MEMORY_RELIABLE
 	/* skip mirror memory regions with MEMBLOCK_NOMIRROR  */
 	if ((flags & MEMBLOCK_NOMIRROR) && memblock_is_mirror(m))
 		return true;
+#endif
 
 	/* skip nomap memory unless we were asked for it explicitly */
 	if (!(flags & MEMBLOCK_NOMAP) && memblock_is_nomap(m))
@@ -1390,6 +1392,7 @@ phys_addr_t __init memblock_alloc_range_nid(phys_addr_t size,
 	return found;
 }
 
+#ifdef CONFIG_MEMORY_RELIABLE
 /**
  * memblock_alloc_range_nid - allocate boot memory block
  * @size: size of memory block to be allocated in bytes
@@ -1457,6 +1460,7 @@ static phys_addr_t __init memblock_alloc_range_nid_flags(
 
 	return found;
 }
+#endif
 
 /**
  * memblock_phys_alloc_range - allocate a memory block inside specified range
@@ -1613,6 +1617,7 @@ void * __init memblock_alloc_try_nid_raw(
 				       false);
 }
 
+#ifdef CONFIG_MEMORY_RELIABLE
 void * __init memblock_alloc_try_nid_raw_flags(
 			phys_addr_t size, phys_addr_t align,
 			phys_addr_t min_addr, phys_addr_t max_addr,
@@ -1645,6 +1650,7 @@ void * __init memblock_alloc_try_nid_raw_flags(
 
 	return ptr;
 }
+#endif
 
 /**
  * memblock_alloc_try_nid - allocate boot memory block
diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index 84acad01ff8a..15fbf83846e0 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -3851,9 +3851,11 @@ get_page_from_freelist(gfp_t gfp_mask, unsigned int order, int alloc_flags,
 		struct page *page;
 		unsigned long mark;
 
+#ifdef CONFIG_MEMORY_RELIABLE
 		/* skip non-movable zone for normal user tasks */
 		if (skip_none_movable_zone(gfp_mask, z))
 			continue;
+#endif
 
 		/*
 		 * CDM nodes get skipped if the requested gfp flag
@@ -5240,6 +5242,7 @@ unsigned long __alloc_pages_bulk(gfp_t gfp, int preferred_nid,
 }
 EXPORT_SYMBOL_GPL(__alloc_pages_bulk);
 
+#ifdef CONFIG_MEMORY_RELIABLE
 static inline void prepare_before_alloc(gfp_t *gfp_mask)
 {
 	bool zone_movable;
@@ -5349,6 +5352,16 @@ static inline bool check_after_alloc(gfp_t *gfp, unsigned int order,
 out:
 	return false;
 }
+#else
+static inline void prepare_before_alloc(gfp_t *gfp_mask) {}
+static inline bool check_after_alloc(gfp_t *gfp, unsigned int order,
+				     int preferred_nid,
+				     struct alloc_context *ac,
+				     struct page **_page)
+{
+	return false;
+}
+#endif
 
 /*
  * This is the 'heart' of the zoned buddy allocator.
diff --git a/mm/shmem.c b/mm/shmem.c
index 258179b89a9b..a15f661a07dd 100644
--- a/mm/shmem.c
+++ b/mm/shmem.c
@@ -1561,22 +1561,6 @@ static struct page *shmem_alloc_page(gfp_t gfp,
 	return page;
 }
 
-static inline bool shmem_prepare_alloc(gfp_t *gfp_mask)
-{
-	if (!shmem_reliable_is_enabled())
-		return true;
-
-	if (mem_reliable_shmem_limit_check()) {
-		*gfp_mask |= GFP_RELIABLE;
-		return true;
-	}
-
-	if (reliable_allow_fb_enabled())
-		return true;
-
-	return false;
-}
-
 static struct page *shmem_alloc_and_acct_page(gfp_t gfp,
 		struct inode *inode,
 		pgoff_t index, bool huge, int node_id)
-- 
2.17.1

